content/blog/client-futures.md

```diff
@@ -0,0 +1,72 @@
+
+
+## Pin
+
+Futures 0.3 makes heavy use of the [`Pin`](https://doc.rust-lang.org/nightly/std/pin/index.html) type, in particular for the type of `self` in the signature of `Future::poll`. Other than the mechanical change to signatures, I had to do a bit of juggling using `Pin::get_unchecked_mut` and `Pin::new_unchecked` (both unsafe) to project fields of futures.
+
+Pinning is a subtle and complicated concept. I still don't feel like I have a great grasp of it all. The best reference I'm aware of are the [std::pin docs](https://doc.rust-lang.org/nightly/std/pin/index.html). I'll try and summarise the ideas here (there are some important subtleties that I'm not going to touch on at all, this is not intended to be a tutorial on pinning):
+
+* `Pin` is a type constructor which only makes sense when applied to a pointer type, e.g., `Pin<Box<_>>`.
+* `Pin` itself is a marker/wrapper type (a bit like [`NonNull`](https://doc.rust-lang.org/nightly/std/ptr/struct.NonNull.html)), not a pointer type. You should read `Pin<Box<Foo>>` as 'a pinned Box pointer to Foo' or 'a Box<Foo> which is guaranteed to be pinned'.
+* If a pointer is pinned, that means that the value pointed to by the pointer will never be moved (moving happens when a non-`Copy` object is passed by value, or via `mem::swap`). Note that the value is allowed to move *before* the pointer is pinned, but not *after*.
+* If a type implements the `Unpin` trait, it means that it never actually matters whether that type is moved or not. I.e., we can safely pretend that a pointer to that type is pinned, even when it is not.
+* `Pin` and `Unpin` are not built into the language at all (although some language features indirectly rely on pinning guarantees). Pinning is compiler-enforced without the compiler knowing anything about it. (This is really cool, and is a great demonstration of how powerful Rust's trait system is for this sort of thing). It works like this: `Pin<P<T>>` only permits safe access to `P`, which does not permit moving any value that `P` points to, unless `T` implements `Unpin` (i.e., the programmer has asserted that `T` does not care about being moved). Operations which could allow a non-`Unpin` value to be moved (basically mutable access) is `unsafe`, and it is up to the programmer to not move anything and guarantee that nothing can be moved later in safe code.
+
+OK, back to the futures migration. Whenever you end up using unsafe methods on `Pin` you have to think about all of the above in order to ensure the pinning invariants. See the [std::pin docs](https://doc.rust-lang.org/nightly/std/pin/index.html) for more on how to reason about this. In many of the places I had to use field projection, the reason is to call `poll` on another future (sometimes indirectly). For that you need a pinned pointer and so you need *structural pinning* (i.e., you project a field of `Pin<&mut T>` to `Pin<&mut FieldType>`).
```

Comment:
```suggestion
OK, back to the futures migration. Whenever you end up using unsafe methods on `Pin`, you have to think about all of the above in order to ensure the pinning invariants. See the [std::pin docs](https://doc.rust-lang.org/nightly/std/pin/index.html) for more details on how to reason about this. In many places, I had to use field projection so that I could call `poll` on another future (sometimes indirectly). To call `poll`, you need a pinned pointer and so you need *structural pinning* (i.e., you project a field of `Pin<&mut T>` to `Pin<&mut FieldType>`).
```